### PR TITLE
Merge branch 'asmexpand' of github.com:AbsInt/CompCert

### PR DESCRIPTION
**Pull request by bschommer**
New pull request for asmexpand:
This branch moves the builtin functions of the arm and ia32 backend into their corresponding Asmexpand files like it is done for the PowerPC backend.

### CODE REVIEW

File: arm/Asmexpand.ml [hunk 1]
```diff
@@ -11,8 +11,334 @@
...
+let expand_builtin_memcpy_big sz al src dst =
+  assert (sz >= al);
+  assert (sz mod al = 0);
+  assert (src = IR2);
+  assert (dst = IR3);
+  let (load, store, chunksize) =
+    if al >= 4 then (Pldr (IR12,src,SOimm _4), Pstr (IR12,dst,SOimm _4) , 4)
+    else if al = 2 then (Pldrh (IR12,src,SOimm _2), Pstrh (IR12,dst,SOimm _2), 2)
+    else (Pldrb (IR12,src,SOimm _1), Pstrb (IR12,dst,SOimm _1), 1) in
+  expand_movimm IR14 (coqint_of_camlint (Int32.of_int (sz / chunksize)));
```
Comment (xavierleroy) [inline arm/Asmexpand.ml]: The original code was using post-indexed addressing, where  the index register is incremented after the memory access.  Pldr* and Pstr* implement offset addressing.  The code generated here does not implement memcpy.

File: arm/Asmexpand.ml [hunk 2]
```diff
@@ -11,8 +11,334 @@
...
+(* The ARM has strict alignment constraints for 2 and 4 byte accesses.
+   8-byte accesses must be 4-aligned. *)
+
+let expand_builtin_memcpy_small sz al src dst =
+  let rec copy ofs sz  =
+    if sz >= 8 && al >= 4 && !Clflags.option_ffpu then begin
+	emit (Pfldd (FR7,src,ofs));
+	emit (Pfstd (FR7,dst,ofs));
+	copy (Int.add ofs _8) (sz - 8)
+      end else if sz >= 4 && al >= 4 then begin
```
Comment (xavierleroy) [inline arm/Asmexpand.ml]: Here and elsewhere I'd prefer the "end" to be indented like the "if", not one space to the right.

File: arm/Asmexpand.ml [hunk 3]
```diff
@@ -11,8 +11,334 @@
...
+  | "__builtin_clz", [IR a1], [IR res] ->
+     emit (Pclz (IR res, IR a1))
+  (* Float arithmetic *)
+  | "__builtin_fabs",  [FR a1], [FR res] ->
+     emit (Pfabsd (res,a1))
+  | "__builtin_fsqrt", [FR a1], [FR res] ->
+     emit (Pfsqrt (res,a1))
+  (* 64-bit integer arithmetic *)
+  | "__builtin_negl", [IR ah;IR al], [IR rh; IR rl] ->
+     emit (Prsbs (rl,al,SOimm _0));
```
Comment (xavierleroy) [inline arm/Asmexpand.ml]: I believe one needs to go through "expand_int64_arith" here, just like the original code goes through "print_int64_arith", so as to handle the case rl = ah

File: arm/Asmexpand.ml [hunk 4]
```diff
@@ -11,8 +11,334 @@
...
+     if !Clflags.option_mthumb then begin
+	 emit (Pmvn (rh,SOreg ah));
```
Comment (xavierleroy) [inline arm/Asmexpand.ml]: There was a comment here in the original code that explains the magic and should be left.

File: arm/TargetPrinter.ml [hunk 5]
```diff
@@ -716,11 +475,15 @@ module Target (Opt: PRINTER_OPTIONS) : TARGET =
       | SOror(r, n) -> fprintf oc "%a, ror #%a" ireg r coqint n
 
     let print_instruction oc = function
-        (* Core instructions *)
+      (* Core instructions *)
+      | Padc (r1,r2,so) ->
+	 fprintf oc "	adc	%a, %a, %a\n" ireg r1 ireg r2 shift_op so; 1
       | Padd(r1, r2, so) ->
           fprintf oc "	add%s	%a, %a, %a\n"
             (if !Clflags.option_mthumb && r2 <> IR14 then "s" else "")
```
Comment (xavierleroy) [inline arm/TargetPrinter.ml]: Excellent idea to special-case r2 = IR14 here.

File: powerpc/Asmexpand.ml [hunk 6]
```diff
@@ -71,6 +33,14 @@ let _8 = coqint_of_camlint 8l
 let _m4 = coqint_of_camlint (-4l)
 let _m8 = coqint_of_camlint (-8l)
 
+let emit_loadimm r n =
+  List.iter emit (Asmgen.loadimm r n [])
+
+let emit_addimm rd rs n =
+  List.iter emit (Asmgen.addimm rd rs n [])
+
```
Comment (xavierleroy) [inline powerpc/Asmexpand.ml]: Are these functions (emit_loadimm, emit_addimm) used anywhere?

Comment (bschommer) [inline powerpc/Asmexpand.ml]: They are used in some of the functions, e.g. expand_builtin_memcpy, etc.

**Comment by xavierleroy** [comment]
Apologies for sleeping on this PR, I didn't notice the updates on Jul 14.  The changes look good and I'm merging them.

**Pull request merged**